torch/csrc/jit/script/module.cpp

```diff
@@ -137,6 +122,249 @@ void Module::to_impl(
+  //     subgraph->inputs().at(i)->replaceAllUsesWith(
+  //         subgraph->inputs().at(it->second));
+  //     subgraph->eraseInput(i);
+  //   }
+  // }
+
+  return std::make_pair(std::move(g), std::move(extra_ivalues));
+}
+
+Method& Module::lower_first_class_method(Function* fn) {
+  fn->ensure_defined();
+  auto lowered = lower_graph(module_object(), *fn->graph());
+  Function& new_func =
+      lowered_methods_.create_function(fn->name(), lowered.first);
+
+  // generate the new schema
+  // slice away the self argument
+  std::vector<Argument> args(
+      fn->getSchema().arguments().begin() + 1,
+      fn->getSchema().arguments().end());
+  size_t id = 0;
+  for (const Slot& slot : lowered.second) {
+    std::ostringstream ss;
+    ss << "slot" << id++;
+    args.emplace_back(ss.str(), slot.type());
+  }
+  new_func.setSchema(fn->getSchema().withArguments(std::move(args)));
+  return _create_lowered_method(&new_func, std::move(lowered.second));
+}
+
+static void createFirstClassValues(
+    Module* module,
+    Value* self,
+    std::unordered_map<Slot, Value*>& result) {
+  auto& g = *self->owningGraph();
+
+  std::vector<Node*> created;
+  struct ToScan {
+    Module* mod;
+    Value* v; // value representing module in the graph
+  };
+  std::vector<ToScan> to_scan = {{module, self}};
+
+  while (!to_scan.empty()) {
+    auto s = to_scan.back();
+    to_scan.pop_back();
+    size_t offset = 0;
+    for (const std::string& name :
+         s.mod->module_object()->type()->attributeNames()) {
+      Value* v = g.insertGetAttr(s.v, name);
+      result[Slot(s.mod->module_object(), offset++)] = v;
+      if (std::shared_ptr<Module> sub = s.mod->find_module(name)) {
+        to_scan.emplace_back(ToScan{sub.get(), v});
+      }
+    }
+  }
+}
+
+void Module::lift_lowered_method(Method& m) {
+  auto graph = m.graph()->copy();
+  Value* self = graph->insertInput(0, "self")->setType(module_object()->type());
+  std::unordered_map<Slot, Value*> slot_to_value;
+  if (!m.initial_ivalues().empty()) {
+    WithInsertPoint guard(*graph->nodes().begin());
+    createFirstClassValues(this, self, slot_to_value);
+  }
+
+  size_t orig_graph_inputs_size = graph->inputs().size();
+  for (size_t i = 0; i < m.initial_ivalues().size(); ++i) {
+    size_t input_offset = orig_graph_inputs_size - i - 1;
+    size_t ivalue_offset = m.initial_ivalues().size() - i - 1;
+    graph->inputs()
+        .at(input_offset)
+        ->replaceAllUsesWith(
+            slot_to_value.at(m.initial_ivalues().at(ivalue_offset)));
+    graph->eraseInput(input_offset);
+  }
+
+  if (!m.initial_ivalues().empty()) {
+    // we added _all_ the submodules as first-class values but maybe did not use
+    // them. So remove any dead attribute lookups
+    EliminateDeadCode(graph);
+  }
+
+  Function& new_fn = class_cu().create_function(m.name(), std::move(graph));
+  // created lifted schema
+  std::vector<Argument> new_args = {Argument("_self", module_object()->type())};
```

Comment:
why the preceding `_`?

Reply:
The user might have a function like `def foo(a, self)` and then use keyword argument calling on it. I renamed to $self to make this case impossible and added a comment. This code is short lived because lift_lowered_methods will be removed once we can expose Function directly.